### PR TITLE
Read entire resp body into buffer for retries

### PR DESCRIPTION
**Pull request by willhug**
Summary: Depending on the transport, we might have cases where
the "success" of a request is not determined until the io.ReadCloser
body of the Response has been fully read (example being tchannel
requests that have more than 1 frame (1<<16 bytes)).  If we pass
back the body to the handler and reading the body fails, we no longer
have the ability to retry the request.

This PR reads the response body into a "buffercloser" type that is
a pooled buffer that will return itself to the sync pool when the
"Close" method is called. It also supports "ReadFrom" for wrapping
the incoming io.Reader, and io.WriteTo to allow io.Copy without
allocations.

Test Plan: added tests, have tested this in the frontcar repo with
its megasuite integration tests.

### CODE REVIEW

File: internal/ioutil/buffercloser.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	New: func() interface{} {
+		return &BufferCloser{
+			b: &bytes.Buffer{},
+		}
+	},
+}
+
+// BufferCloser is a light wrapping around the bytes.Buffer that implements
+// a "Close" method to return the buffer to a sync.Pool of buffers.
+type BufferCloser struct {
```
Comment (peter-edge) [inline internal/ioutil/buffercloser.go]: This is another situation where you have a public struct that can be created without using the constructor. If someone does `bufferCloser := &ioutil.BufferCloser{}`, `b` will be nil which means `ReadFrom, WriteTo, Read` will panic, and even if `bytes.Buffer` was not a pointer, the value would not be initialized from the pool anyways, making `Close` not work as intended.

You can fix this in one of two ways:

- For every call to a function on `*BufferCloser`, check if `b` is properly initialized. This is expensive.
- Make `BufferCloser` a private struct, and instead return an interface that implements `io.ReaderFrom, io.WriterTo, io.Closer`. This is what I would recommend, I think the benefits outweigh the compatibility costs of using a struct.

All code in the Golang stdlib that exposes public structs ensures that there will not be issues regardless of if an user creates the struct by declaring it, or with the corresponding constructor function.

Comment (abhinav) [reply internal/ioutil/buffercloser.go]: Exposing public structs, even if the zero-value is not valid, is a very common
pattern used throughout YARPC, other libraries, and the standard library. The
expectation is that if a constructor is provided, unless otherwise documented,
you must use the constructor.

In the standard library at least, the documentation explicitly states that the
zero-value is valid if that's the case rather than letting you assume it.

Some structs in the standard library go out of their way to ensure that they
don't explode in the zero-value case (regexp.Regexp is definitely not one of
them) but I don't think it's worth the effort or performance cost here.

Comment (peter-edge) [reply internal/ioutil/buffercloser.go]: I can't think of an example in the stdlib that does not allow you to use a public struct you just declare, they go out of their way to make sure they work no matter how you declare them, as they should - you're knowingly publicly exposing something that could panic otherwise. If you can think of an example that documents the zero-value is invalid, I'd love to see it, as I've seen comments in the stdlib that say things like "we do this because if someone declares the struct instead of using the constructor it doesn't work".

tl;dr I don't know of an example in stdlib that will panic in this case, and regardless it's a bad pattern to expose something that can error. Hence why it's usually nicer to expose an interface here, or to handle the error cases.

Comment (abhinav) [reply internal/ioutil/buffercloser.go]: > If you can think of an example that documents the zero-value is invalid [..]

Actually, the opposite. Structs explicitly state that their zero value is valid. So you could say that the default is invalid unless stated otherwise.

Simple example of a struct that's invalid for zero: https://golang.org/pkg/regexp/#Regexp. Panic demonstration: https://play.golang.org/p/NQg1akM6iS.

Comment (peter-edge) [reply internal/ioutil/buffercloser.go]: `Regexp` is a special case as how they are constructed is unusual, and I wouldn't agree with that anyways. `bytes.Buffer` is a much more normal case here, as is `http.Server`, etc. All of these work to make sure the zero-value works, as they should.

Comment (abhinav) [reply internal/ioutil/buffercloser.go]: More examples:

- json.Decoder: https://play.golang.org/p/YagPipX545
- log.Logger: https://play.golang.org/p/EOdlj5nWeL
- reflect.Value is all panics but that's probably a special case

I agree that zero-value-just-works is a nice to have but when there is *some* required input, you can't have it. Again, [Accept interfaces return structs](http://idiomaticgo.com/post/best-practice/accept-interfaces-return-structs/) is pretty idiomatic to Go, including when the returned struct doesn't have a valid zero-value.

Comment (peter-edge) [reply internal/ioutil/buffercloser.go]: Yea I get it, ok let's discuss at standup because my personal opinion is that it's really not good to expose something that can panic with incorrect usage (at the very least, `BufferCloser` should document this in the Godoc I'd say right?), I want us to be consistent as a team going forward.

Comment (peter-edge) [reply internal/ioutil/buffercloser.go]: Actually just going to add this to weekly agenda for Thursday, let's discuss then, unblocking this for now.

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,8 @@ Releases
 v1.16.0-dev (unreleased)
 --------------------
 
+- x/retry: Fix bug where large TChannel responses would cause errors in retries.
+
```
Comment (abhinav) [inline CHANGELOG.md]: nit: two empty lines between sections

File: internal/ioutil/buffercloser.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+}
+
+// WriteTo implements io.WriterTo
+func (b *BufferCloser) WriteTo(w io.Writer) (n int64, err error) {
+	return b.b.WriteTo(w)
+}
+
+// Read implements io.Reader
+func (b *BufferCloser) Read(p []byte) (n int, err error) {
+	return b.b.Read(p)
```
Comment (abhinav) [inline internal/ioutil/buffercloser.go]: optional: What's stopping you from embedding `*bytes.Buffer`?

```lang=go
type BufferCloser struct{ *bytes.Buffer }

func NewBufferCloser() *BufferCloser {
  // ...
}

func (b *BufferCloser) Close() error {
  _pool.Put(b.Buffer)
  return nil
}
```

It gives `BufferCloser` other methods like `WriteString` 
(which is used by `io.WriteString` if present).

File: internal/ioutil/buffercloser.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+// BufferCloser is a light wrapping around the bytes.Buffer that implements
+// a "Close" method to return the buffer to a sync.Pool of buffers.
+type BufferCloser struct {
+	b *bytes.Buffer
+}
+
+// NewBufferCloser returns a new BufferCloser from the Buffer pool that has been
+// reset.
+func NewBufferCloser() *BufferCloser {
```
Comment (abhinav) [inline internal/ioutil/buffercloser.go]: FYI you can (but don't have to) pass BufferCloser around by value since its only field is a pointer.

File: internal/ioutil/buffercloser_test.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+func TestBuffers(t *testing.T) {
+	var wg sync.WaitGroup
+	for g := 0; g < 10; g++ {
+		wg.Add(1)
+		go func() {
+			for i := 0; i < 100; i++ {
+				buf := NewBufferCloser()
+				b := make([]byte, 1)
+				n, readErr := buf.Read(b)
```
Comment (abhinav) [inline internal/ioutil/buffercloser_test.go]: nit: why is this called readErr instead of err

File: internal/ioutil/buffercloser.go (hunk 3 above)
Comment (kriskowal) [reply internal/ioutil/buffercloser.go]: Could technically just `type BufferCloser bytes.Buffer` and implement `*BufferCloser` receivers.

File: x/retry/retry.go [hunk 6]
```diff
@@ -199,3 +205,16 @@ func isIdempotentProcedureRetryable(err error) bool {
 		return false
 	}
 }
+
+func readBody(src io.ReadCloser) (io.ReadCloser, error) {
+	buffer := ioutil.NewBufferCloser()
+	if _, err := buffer.ReadFrom(src); err != nil {
+		buffer.Close() // Make sure we cleanup the buffer.
```
Comment (kriskowal) [inline x/retry/retry.go]: nit: "clean up" verb "cleanup" noun.

File: internal/yarpctest/outboundtest/outbound_event.go [hunk 7]
```diff
@@ -164,3 +171,25 @@ func assertEqualIfSet(t require.TestingT, want, got string, msgAndArgs ...interf
 		assert.Equal(t, want, got, msgAndArgs...)
 	}
 }
+
+type customRespBody struct {
```
Comment (kriskowal) [inline internal/yarpctest/outboundtest/outbound_event.go]: nit: At some point we should iron out resp vs res throughout. We are not consistent with our abbreviations. I favor Res because it’s pronounceable and same length as Req.

File: internal/ioutil/buffercloser_test.go [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	"bytes"
+	"io"
+	"math/rand"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestBuffers(t *testing.T) {
```
Comment (kriskowal) [inline internal/ioutil/buffercloser_test.go]: Consider documenting the intended methodology here. From reading, it looks like the intent is to exercise the buffer pool by using two BufferClosers alternately take from the shared pool.
